src/main/java/org/alfresco/repo/security/authentication/identityservice/IdentityServiceAuthenticationComponent.java

```diff
@@ -0,0 +1,104 @@
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Alfresco. If not, see <http://www.gnu.org/licenses/>.
+ * #L%
+ */
+package org.alfresco.repo.security.authentication.identityservice;
+
+import org.alfresco.repo.management.subsystems.ActivateableBean;
+import org.alfresco.repo.security.authentication.AbstractAuthenticationComponent;
+import org.alfresco.repo.security.authentication.AuthenticationException;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.keycloak.authorization.client.AuthzClient;
+import org.keycloak.authorization.client.util.HttpResponseException;
+
+/**
+ *
+ * Authenticates a user against Keycloak.
+ * Keycloak's {@link AuthzClient} is used to retrieve an access token for the provided user credentials,
+ * user is set as the current user if the user's access token can be obtained.
+ * <br>
+ * The AuthzClient can be null in which case this authenticator will just fall through to the next one in the chain.
+ *
+ */
+public class IdentityServiceAuthenticationComponent extends AbstractAuthenticationComponent implements ActivateableBean
+{
+    private final Log logger = LogFactory.getLog(IdentityServiceAuthenticationComponent.class);
+    private AuthzClient authzClient;
+    private boolean active;
+
+    public void setAuthenticatorAuthzClient(AuthzClient authenticatorAuthzClient)
+    {
+        this.authzClient = authenticatorAuthzClient;
+    }
+
+    public void authenticateImpl(String userName, char[] password) throws AuthenticationException
+    {
+
+        if (authzClient == null)
+        {
+            if (logger.isDebugEnabled())
+            {
+                logger.debug("AuthzClient was not set, possibly due to the 'identity-service.disable-username-password-authentication=true' property. ");
+            }
+
+            throw new AuthenticationException("User not authenticated because AuthzClient was not set.");
+        }
+
+        try
+        {
+            // Attempt to get an access token using the user credentials
+            authzClient.obtainAccessToken(userName, new String(password));
+
+            // Successfully obtained access token so treat as authenticated user
+            setCurrentUser(userName);
+        }
+        catch (HttpResponseException e)
+        {
+            if (logger.isDebugEnabled())
+            {
+                logger.debug("Failed to authenticate user against Keycloak. Status: " + e.getStatusCode() + " Reason: "+ e.getReasonPhrase());
+            }
+
+            throw new AuthenticationException("Failed to authenticate user against Keycloak.", e);
+        }
+    }
+
+    public void setActive(boolean active)
+    {
+        this.active = active;
+    }
+
+    @Override
+    public boolean isActive()
+    {
+        return active;
+    }
+
+    @Override
+    protected boolean implementationAllowsGuestLogin()
+    {
+        return true;

```

Comment:
We might need to have this configurable.
Guest logins are configurable in other auth subsystems.

Reply:
I think it would still be configurable as the abstract class has the same flag which takes precedence over the implementation flag. Not sure why that this is actually there but I added it anyway in case it is used somewhere. I'v also added it into the subsystem properties.